### PR TITLE
Add BPE dropout support, use it in training.

### PR DESCRIPTION
**Pull request by howard0su**

### CODE REVIEW

**Comment by howard0su** [comment]
I don't have hardware to run through training and validation. I just noticed this when I was reading BPE code for #2023

File: llama.cpp [hunk 1]
```diff
@@ -1814,13 +1818,26 @@ struct llama_tokenizer {
         work_queue_.push(bigram);
     }
 
+    bool skip_merge()
+    {
+        std::uniform_real_distribution<> gen(0.0, 1.0);
+        if (dropout_ <= 0.0) {
+            return false;
+        }
+        if (dropout_ >= 1.0)
```
Comment (Green-Sky) [inline llama.cpp:1827]: please add block brackets

Comment (Green-Sky) [reply llama.cpp:1827]: also, why would you specify > 100% dropout?

File: llama.cpp [hunk 2]
```diff
@@ -1814,13 +1818,26 @@ struct llama_tokenizer {
...
+            return true;
+        return gen(rng) < dropout_;
+    }
+
     const llama_vocab & vocab_;
     std::vector<llama_sp_symbol> symbols_;
     llama_sp_bigram::queue work_queue_;
+    float dropout_;
+    std::mt19937 rng;
```
Comment (Green-Sky) [inline llama.cpp:1836]: at some point we should replace all the mersenne twisters ...

File: llama.cpp (hunk 1 above)
Comment (howard0su) [reply llama.cpp:1827]: i believe it is used for input validation and protection. Copied from https://github.com/google/sentencepiece/blob/master/src/bpe_model.cc#L132-L139

**Review by Green-Sky — APPROVED** [review]

**Comment by howard0su** [comment]
Is it okay for me to add new parameter to tokenize API?

**Comment by alitariq4589** [comment]
Can one of the admins verify this patch?

**Pull request closed without merging**